src/tdigest.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+/*
+ * Interface for t-digest support
+ *
+ * Originally from PipelineDB 1.0.0
+ * Original copyright (c) 2018, PipelineDB, Inc. and released under Apache License 2.0
+ * Modifications copyright by Timescale, Inc. per NOTICE
+ */
+
+#ifndef TDIGEST_H
+#define TDIGEST_H
+
+#include "c.h"
+#include "nodes/pg_list.h"
+
+#define TDIGEST_DEFAULT_COMPRESSION 200
+#define TDIGEST_MIN_COMPRESSION 20
+#define TDIGEST_MAX_COMPRESSION 1000
+
+/*
+ * A centroid is the fundamental unit of t-digest, and it represents an
+ * aggregate data values by storing the mean of the values and how many
+ * values contribute to that mean.
+ */
+typedef struct Centroid
+{
+	// number of samples contributing to this centroid's mean
+	uint64 weight;
+	// the average value of the samples represented by this centroid
+	float8 mean;
+} Centroid;
+
+/*
+ * A t-digest structure holds all the centroids that represent a given
+ * dataset, as well as a few extra values for use in maintaining the data
+ * structure effciently.
+ */
+typedef struct TDigest
+{
+	uint32 vl_len_; // for postgres base type definition
+
+	float8 compression; // increasing compressing lowers memory usage and decreases accuracy
+	uint32 threshold;   // maximum number of centroids to store before automatically compressing
+	uint32 size;		// max number of centroids that can be stored in the t-digest
+
+	int64 total_weight; // total num of merged samples represented by the t-digest
+	float8 min;			// lowest centroid mean in the t-digest
+	float8 max;			// highest centroid mean in the t-digest
+
+	List *unmerged_centroids; // input buffer for new centroids
```

Comment:
The combination of this `List` along with the `vl_len_` is very suspicious; either it cannot be stored on disk, and thus is an `internal` type and does not need the header, or it can be, in which case there's a weird invariant that `unmerged_centroids` should always be read/written as `NIL`. If it's the latter, then the `List` does not belong in this `struct`; it should either be a seperate struct, as in an "expanded" view, or it should be a array laid out directly after this header struct in memory.